### PR TITLE
Regression with categorical data and introducing formula language

### PR DESCRIPTION
**Pull request by lokeshh**
Introduce formula language for regression and add support for category variables.

### CODE REVIEW

**Comment by lokeshh** [comment]
@v0dro , @agarie , @zverok Should I write the specs in Rspec or in MiniTest which Statsample currently uses? If in the future we are going to turn to Rspec from MiniTest then its better to write the tests in Rspec now.

File: Rakefile [hunk 1]
```diff
@@ -1,4 +1,5 @@
 $:.unshift File.expand_path("../lib/", __FILE__)
+lib_folder = File.expand_path("../lib", __FILE__)
```
Comment (lokeshh) [inline Rakefile]: Perhaps line 2 is not required. Any suggestions?

File: test/test_fit_model.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  end
+  context '#df_for_regression' do
+    it 'gives correct dataframe when no interaction' do
+      @formula = 'y~a+e'
+      @vectors = %w[a e_B e_C y]
+    
+      @model = Statsample::FitModel.new @formula, @df
+
+      @model.df_for_regression.vectors.to_a.sort.must_equal @vectors.sort
+    end
```
Comment (lokeshh) [inline test/test_fit_model.rb]: It doesn't seem to work. Giving me the following error:

```ruby
  1) Error:
Statsample::FitModel#test_0001_gives correct dataframe when no interaction:
ArgumentError: comparison of Symbol with String failed
    test/test_fit_model.rb:16:in `sort'
    test/test_fit_model.rb:16:in `block (3 levels) in <main>'

1 runs, 0 assertions, 0 failures, 1 errors, 0 skips
```

**Comment by zverok** [comment]
BTW, as far as I understand, newest RSpec (3.5.0) has support for MiniTest: you can use RSpec matchers in MiniTest-tested code. Maybe this possibility could be investigated, for gradual and smooth transition.

File: test/test_fit_model.rb (hunk 2 above)
Comment (lokeshh) [inline test/test_fit_model.rb]: Resolved.

**Comment by v0dro** [comment]
@lokeshh since statsample-glm's regression is slow, we'll just stick to statsample for now and think of removing regression from statsample sometime later.

Can you fix this PR so that I can merge it? Let the regression and formula language remain in statsample the way it is now.

**Comment by lokeshh** [comment]
Ok. I will finish it.

**Comment by lokeshh** [comment]
Hmm... its not running the Travis tests at the bottom. Any ideas why?

**Comment by v0dro** [comment]
Did you try restarting the build?

**Comment by lokeshh** [comment]
Its working now. Thanks

**Comment by v0dro** [comment]
That's a weird failure lol. Can you modify the failing test and just introduce some sort of tolerance level?

**Comment by lokeshh** [comment]
This is weird. I just ran the tests in my system with Ruby 2.0 and all tests are passing.

**Comment by lokeshh** [comment]
Tried it on another computer with Ruby 2.2.4, there also it pases all the tests. Let me run the Travis again.

**Comment by lokeshh** [comment]
@v0dro Any idea how I may go ahead with this. The Travis is failing but the same is passing on my and some other system I tried.

**Comment by v0dro** [comment]
Let me have a look hold on.

**Comment by v0dro** [comment]
In `.../statsample/test/test_factor_pa.rb:42`, make a small change to the test.

Instead of having the error tolerance at `0.05`, make it `0.07`. I think that should resolve it.

**Comment by lokeshh** [comment]
Thanks! It worked :+1:

**Comment by lokeshh** [comment]
No wait.

**Comment by lokeshh** [comment]
This can be merged now.

**Pull request merged**